### ISSUE TITLE
Duplicate Detection Performance Investigation

### ISSUE THREAD

**Issue by NoMoreFood** [windirstat/windirstat#156]
@AartBluestoke @linuxgurugamer

Continuing the discussion from https://github.com/windirstat/windirstat/issues/145.  I made some tweaks to the duplicate detection in the latest beta:  https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2024-12-17

It seems to work much better for me now.  I can scan my entire 2 TB C: drive (75% full) in 9 minutes which seems pretty reasonable.

**Comment by linuxgurugamer** [comment]
Downloaded and tested.  It did go much faster, but now it seems to be showing the same file as duplicated 23005 times, others as well, but I think this picture will show my point:

![image](https://github.com/user-attachments/assets/2f0603ef-24eb-4017-8caa-534ad990cac9)

**Comment by linuxgurugamer** [comment]
Another thing:  The following pic shows the dups sorted by items.  The first one has 611,016 dups, when I click on the + sign, it hangs, I assume because it's trying to add all 611016 items to the list box.  You should limit the number of items added to a reasonable number:

![image](https://github.com/user-attachments/assets/28c4709b-ea73-4312-abef-5c97dd5eabc9)

**Comment by NoMoreFood** [comment]
Ok, I'll take a look, thanks!

**Comment by reukiodo** [comment]
Would it be faster to focus on the largest files first? ie: sort by size, then try duplicate detection in some size ranges ? Thousands of small files duplicated (0 byte files) I can handle, but even a few duplicates of multi-hundered-gb files are more concerning.

**Comment by reukiodo** [comment]
Even better than just deletes is if WinDirStat can handle replacing the duplicates with hardlinks if the data is indeed all the same and on the same filesystem.

**Comment by linuxgurugamer** [comment]
> Would it be faster to focus on the largest files first? ie: sort by size, then try duplicate detection in some size ranges ? Thousands of small files duplicated (0 byte files) I can handle, but even a few duplicates of multi-hundered-gb files are more concerning.

I'd agree, except that for what I showed, the paths were identical.  That shouldn't happen


> Even better than just deletes is if WinDirStat can handle replacing the duplicates with hardlinks if the data is indeed all the same and on the same filesystem.

Some other dup finders actually do that.  Makes a lot of sense, if it can be done.

**Comment by NoMoreFood** [comment]
@linuxgurugamer Can you still see if you have the issue with the repeated hashes in this version:

https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2024-12-18

Thanks!

**Comment by linuxgurugamer** [comment]
> @linuxgurugamer Can you still see if you have the issue with the repeated hashes in this version:
> 
> https://github.com/windirstat/windirstat/releases/tag/beta/v2.1.2/2024-12-18
> 
> Thanks!

Yes, it seems to be resolved. Bonus is that it now takes less than a minute to do what the last version took about 5 minutes, and the one before took forever.

A suggestion (see what others think before implemnting):  Maybe put a minimum size to report on?  It found over 4700 zero-length files, which don't take any significant space and I'd be leery of deleting those files, most of them seem to be sqlite files (at least on my system)

**Comment by linuxgurugamer** [comment]
Ok, looking much better, but I ran into a small problem:

I looked at the duplicates, saw an entire directory I could remove, so I did.  Then I clicked the rescan button.  It seemed to hang, I had to kill it in the task manager

**Comment by linuxgurugamer** [comment]
Another suggestion:  When looking at the duplicates, the hash isn't useful for identifying what the file(s) are.  I'd suggest one of the following:

1.  Show the most commonly used file name in all the duplicates, or if none match, show the first
2. Maybe only show duplicates for specified file types

**Comment by NoMoreFood** [comment]
> Ok, looking much better, but I ran into a small problem:
> 
> I looked at the duplicates, saw an entire directory I could remove, so I did. Then I clicked the rescan button. It seemed to hang, I had to kill it in the task manager

Did you do rescan selected or rescan all?  I'll see if I can repro.

**Comment by linuxgurugamer** [comment]
Just tested this.  

I scanned just my Pictures directory (with dups enabled), when it finished I licked the top left button (open a collection of drives), that worked.
Then I clicked the button "Rescan the whole directory tree", that also worked.

I then scanned the entire drive (1.4 Tb, 900,000 files), repeasted the above tests.
The first button, "Open a collection of drives" worked.
The second button, "Rescan the whole directory tree", seemed to hang, According to the Task manager, there were 10 threads running, usiong about 1.1g of memory, and was using about 3% of the cpu.  

After a while, it seemed to finish.  The issue here seems to be, no visual indication that it was running (ie:  no Pacmans)  for a small directory that's fine, but for a large drive, it can be disconcerting

**Comment by AartBluestoke** [comment]
using 2024-12-18 version - a couple of time points - it still seems it goes quicker then slows down:
<img width="639" alt="2024-12-20 15_25_49-WinDirStat - 12% C__D__H_  (Administrator)" src="https://github.com/user-attachments/assets/18dd5ab8-b1df-4426-8e59-b3637706531d" />

<img width="687" alt="2024-12-20 15_29_13-" src="https://github.com/user-attachments/assets/1c8b5350-7dc8-4f44-a957-16001b006fdf" />
<img width="668" alt="image" src="https://github.com/user-attachments/assets/e187719f-aa50-4fa2-acdd-ff82f034709c" />

<img width="637" alt="image" src="https://github.com/user-attachments/assets/acc356d1-e0ed-427f-b85d-1f43f15e5549" />

**Comment by crackwitz** [comment]
I might have an idea to speed things up, depending on how you structure the problem.

The problem is to tell whether files are different.

- Files of different sizes must be different (cannot be the same). If they are the same size... we can't say and have to test further.
- Files with different hashes must be different. If they have the same hash... absent a collision, they *are* the same file.

(not going into inodes that have multiple links, but that's a sensible thing to respect in any deduplication effort)

I'm assuming here that the file size test already happens in WDS. That trivially excludes most big files from having to be read at all, under the assumption that each such file has a fairly unique file size.

Still, that leaves many files that have "coarse" file sizes, e.g. multiples of some block size.

You can insert a third test between the file size (nothing to read) and hashing (whole file to read) that reads *a part of* the file. There could be all kinds of schemes. In the past, I chose a *sparse* hash called [UUHash](https://en.wikipedia.org/wiki/UUHash) that requires about O(log n) bytes of the file to be read. It's not comparable to a cryptographic hash of the entire file but it doesn't have to be.

I implemented that approach in a Python script I used to curate a particular hoard of data. I didn't track how much that spared me from reading, compared to just size and hash over entire file. It felt like a lot. I'd have to look back into my code for details. It was basically hashmaps to index sets of files by size/sparse hash/dense hash.

**Comment by reukiodo** [comment]
In addition, as a step even before comparing file sizes, is to just throw out (exclude) any 0-sized file.

**Comment by NoMoreFood** [comment]
Released with https://github.com/windirstat/windirstat/releases/tag/release/v2.2.0

If any additional action is warranted, please submit a new issue.
